[Experimental] Product filters/active block new structure

roykho opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/51641

This PR refactors the active filter to the new design structure. Here are some additional notes.

- Fixed clear button setting where toggling it was not working in the editor/frontend.
- Changed the design of the chips to the new spec.
- Remove list view setting as we're just wanting to only chips.
- I did not implement the editor active chips (real data) in this PR. I think we can address this separately because one idea is we could actually use the same data from Attributes filter. So instead of calling the API multiple times, we call it once in the parent Product Filters block and pass the data through context to the child filters.
- Note that I found a bug where the clear button has to be above the filter list/chips block. This bug exists with other filters blocks as well and not specific to this block. But it doesn't block this PR. We can fix that bug in another PR. See step 7 to get around this bug and test this PR.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

> [!NOTE]  
> **Pre-requisite**
> - If you're using WC Beta Tester plugin, make sure the `experimental-blocks` feature is enabled (Tools > WCA Test Helper > Features).
> - Make sure to reset Product Catalog Template.

1. Navigate to Product Catalog template (Appearance > Editor > Templates > Product Catalog template.
2. Edit the template.
3. With the block inserter, add Product Filters (Experimental) block to the template.
4. Open the Document Overview (List view) and ensure you see the Product Filter Active block in the new design structure.
5. Select the parent Active Filters block, check the inspector controls and ensure you see the Clear button setting there.
6. Ensure you can enable/disable the Clear button and have it reflected in the editor.
7. Move the clear button to above the active chips block and save. View in the frontend and ensure the Clear button reflects your saved changes by choosing some filters.
8. Back in the editor Document Overview, ensure you see the Chips block. You should also see placeholder active chips in the editor.
9. Select the Chips block and in the inspector controls you should see option to change colors of the text, border and background. Have a play around with those and ensure they reflect in the editor and in the frontend.
10. In the frontend shop page (assuming you added the filter to the shop page), have a play around with the attributes and ensure the Active filter is working correctly by displaying the selected attributes.
11. Try removing a single active selected attribute by clicking on the X (remove button) on the chip. Ensure it removes the attribute filter and from the URL.
12. Now select more than one attributes and ensure you see the Clear button next to the active filter title.
13. Click on the Clear button and ensure it clears all the filters and in the URL.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

roykho committed:
Update product filters active block to new structure

roykho committed:
Merge branch 'trunk' into product-filters/active-block-new-structure

roykho committed:
Build active chips items

roykho committed:
Add changelog

roykho committed:
Remove unneeded line-height

roykho committed:
Fix linting

roykho committed:
Fix e2e tests

roykho committed:
Refactor removeText

roykho committed:
Clean up

roykho committed:
Add clear button even without group

roykho committed:
Merge branch 'trunk' into product-filters/active-block-new-structure

roykho committed:
Fix stock status name in e2e test

roykho committed:
Remove heading block

roykho committed:
Remove unused i18n module

roykho committed:
Merge branch 'trunk' into product-filters/active-block-new-structure

roykho committed:
Remove period

roykho committed:
Add clear button variation description

roykho committed:
Fix typo

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/active-chips/block.json:
What do you think about renaming this to "Removable Chips"? I think it expresses the function of the block better than "Active Chips".

roykho replied:
In the design description the chips block says "Display active filters as chips.". If we changed the Chips name to Removable Chips, it seems contradicting the block description.

Also if we think about what the chips is actually doing is showing the "active" filtered chips right? Sure they're removable but that is not their main function? Main function is to list the chips.

dinhtungdu replied:
> In the design description the chips block says "Display active filters as chips."

Can you share a link to this? I can only see the description for the Active block which is:

> Let shoppers view and manage currently active product filters.

roykho replied:
https://github.com/woocommerce/woocommerce/pull/52268/files#diff-9afed804e854709c1ee5b36bf01d53cd67c79057c2f8822577f8236bbd69dcdeR6

dinhtungdu replied:
> [#52268 (files)](https://github.com/woocommerce/woocommerce/pull/52268/files#diff-9afed804e854709c1ee5b36bf01d53cd67c79057c2f8822577f8236bbd69dcdeR6)

Where does this come from? We should use the one from the new design.

roykho replied:
Can you show me where to find the description for the chips in the new design?

roykho replied:
In the interest of not prolonging this. I changed it to "Removable Chips". I will push up the changes soon.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/clear-button/block-variations.tsx:
Do we need a variation here? IMO displaying the Clear button in the List view doesn't create any confusion. It's also what we have in the design: 
![image](https://github.com/user-attachments/assets/9e292f1c-3b35-4df2-a826-e429b578ec78)

dinhtungdu replied:
You can still change the default text of the button too, using block template.

roykho replied:
Yeah I tried it without the block variation first but I couldn't get it to work to have customized button text and block description. Do you have an example how we could pass the button text 3 levels deep? Because Clear button is just a wrapper for the core/buttons and core/button block.

dinhtungdu replied:
You can customize the text of the button like this:
```js
	const { children, ...innerBlocksProps } = useInnerBlocksProps(
		useBlockProps(),
		{
			allowedBlocks: getAllowedBlocks( EXCLUDED_BLOCKS ),
			template: [
				[
					'woocommerce/product-filter-active-chips',
					{
						lock: {
							remove: true,
						},
					},
				],
				[
					'woocommerce/product-filter-clear-button',
					{
						lock: {
							remove: true,
							move: false,
						},
					},
					[
						[
							'core/buttons',
							{ layout: { type: 'flex' } },
							[
								[
									'core/button',
									{
										text: __( 'Clear all', 'woocommerce' ),
										className:
											'wc-block-product-filter-clear-button is-style-outline',
										style: {
											border: {
												width: '0px',
												style: 'none',
											},
											typography: {
												textDecoration: 'underline',
											},
											outline: 'none',
											fontSize: 'medium',
										},
									},
								],
							],
						],
					],
				],
			],
		}
	);
```

dinhtungdu replied:
But I think we should rework the button (later, in next cycle), and make it just a single block. The current one is quite problematic, for example: the complexity of manipulation here and the hover style.

roykho replied:
What about the description of the block? How do you override that? For ref:
![file.png](https://github.com/user-attachments/assets/25d3547f-6399-4823-9fa9-ef03cbd5ee16)

dinhtungdu replied:
I'm fine with using the same description. Or we can find another one that fits both cases.

roykho replied:
So another words the block description cannot be override correct?

roykho replied:
Wouldn't it just be easier to use block variation in this case?

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterActiveChips.php:
```diff
@@ -0,0 +1,182 @@
+	/**
+	 * Parse the filter parameters from the URL.
+	 * For now we only get the global query params from the URL. In the future,
+	 * we should get the query params based on $query_id.
+	 *
+	 * @param int $query_id Query ID.
+	 * @return array Parsed filter params.
+	 */
+	private function get_filter_query_params( $query_id ) {
+		// phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$request_uri = isset( $_SERVER['REQUEST_URI'] ) ? wp_unslash( $_SERVER['REQUEST_URI'] ) : '';
+
+		$parsed_url = wp_parse_url( esc_url_raw( $request_uri ) );
+
+		if ( empty( $parsed_url['query'] ) ) {
+			return array();
+		}
+
+		parse_str( $parsed_url['query'], $url_query_params );
+
+		/**
+		 * Filters the active filter data provided by filter blocks.
+		 *
+		 * @since 11.7.0
+		 *
+		 * @param array $filter_param_keys The active filters data
+		 * @param array $url_param_keys    The query param parsed from the URL.
+		 *
+		 * @return array Active filters params.
+		 */
+		$filter_param_keys = array_unique( apply_filters( 'collection_filter_query_param_keys', array(), array_keys( $url_query_params ) ) );
+
+		return array_filter(
+			$url_query_params,
+			function ( $key ) use ( $filter_param_keys ) {
+				return in_array( $key, $filter_param_keys, true );
+			},
+			ARRAY_FILTER_USE_KEY
+		);
+	}
```
Why does this block need these methods? The block should only render a list of items with a remove button. All logic for processing the filter options should belong to the Active block.

roykho replied:
So since we decided to create a separate chips block just for active chips, I didn't think it would be reused since it is much more specific to Active filters so I just thought to have the logic live there.

dinhtungdu replied:
If we do this, we break the convention that the display style block is swappable as this block is not simple enough. If we pass the processed list down from the Active block, then Active block can have other display styles (by us or by 3rd parties).
One possible argument here is: that it sounds premature, but we're already doing the same for other filter blocks.

dinhtungdu commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/active-chips/edit.tsx:
```diff
@@ -0,0 +1,183 @@
+							/>
+						</button>
+					</li>
+				</ul>
```
Similar to my comment on the server side of this block, we should pass the list of options from the Active block to this block through block context instead.

dinhtungdu reviewed (changes requested):
Thanks for working on this PR. The block is working as expected in my test. I have some comments, mostly on the active chips block, which is doing too much IMO.

About the clear button, as we discussed in the call, we should have a unified solution of yours here and @thealexandrelara one in https://github.com/woocommerce/woocommerce/pull/52246. We can move your changes (to the Clear button) here to Alexandre PR, what do you think?

roykho committed:
Rename active chips to removable chips

roykho committed:
Extract data items from removable chips to make it reusable

roykho commented:
@dinhtungdu - thanks for review so far. I've pushed up updates.

- Changed the active chips to removable chips
- Extracted the data logic from the removable chips to the parent active filters block in both editor side and frontend server side rendering.
- I've left the block variations for the clear button for now since we said we will tackle it later with a single block instead.

roykho committed:
Fix e2e tests

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterRemovableChips.php:
```diff
@@ -0,0 +1,120 @@
+			},
+			''
+		);
+	}
```
```suggestion
	}

	/**
	 * Get the frontend script handle for this block type.
	 *
	 * @param string $key Data to get, or default to everything.
	 *
	 * @return null
	 */
	protected function get_block_type_script( $key = null ) {
		return null;
	}
```

roykho replied:
> Note that the navigation isn't working with Clear All button enabled. There is a bug in the current run time of Interactivity API that Woo is using (see: p1729680501130659-slack-C04QWL152LX). So let's turn it off by default and create another issue for that problem.

Yeah I mentioned this in the notes.

dinhtungdu reviewed (approved):
Left a minor comment but approving as this is LGTM. Thanks for addressing the feedback!

Note that the navigation isn't working with Clear All button enabled. There is a bug in the current run time of Interactivity API that Woo is using (see: p1729680501130659-slack-C04QWL152LX). So let's turn it off by default and create another issue for that problem.

dinhtungdu reviewed (changes requested):
If I add a block inside the Active block, like Heading, it's visible even if there isn't any active filter.

<img width="1514" alt="image" src="https://github.com/user-attachments/assets/bfa0b245-bd34-48e7-a12d-12d95f355a93">

roykho committed:
Add frontend script loading to null

roykho committed:
Don't render the block if no active filters

roykho committed:
Disable clear button by default

roykho commented:
@dinhtungdu - thanks for the review. I've pushed up updates.

- Don't show active filters block when there are no active filters
-  Made clear button disabled by default.

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/active-filters/index.tsx:
```diff
@@ -10,11 +10,12 @@ import { productFilterActive } from '@woocommerce/icons';
 import metadata from './block.json';
 import Edit from './edit';
-import './style.scss';
+import Save from './save';
```
```suggestion
import save from './save';
```

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/active-filters/index.tsx:
```diff
@@ -10,11 +10,12 @@ import { productFilterActive } from '@woocommerce/icons';
 	registerBlockType( metadata, {
 		icon: productFilterActive,
 		edit: Edit,
+		save: Save,
```
Save is not a component but a function. it shouldn't be in Camelcase.

```suggestion
		save,
```

retrofox commented on plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/active-filters/inspector.tsx:
```diff
@@ -0,0 +1,76 @@
+	}, [ clearButton, clientId ] );
+
+	return (
+		<>
```
unneeded wrapper`<>`

roykho committed:
Clean up

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterActive.php:
```diff
@@ -48,142 +48,37 @@ protected function render( $attributes, $content, $block ) {
-
-		if ( 'chips' === $attributes['displayStyle'] ) {
-			$list_classes .= ' list-chips';
+		if ( empty( $active_filters ) ) {
+			return '';
```
We can't do this as the Removable Chips style won't load initially, see pdnLyh-7qP-p2. We have to render all inner blocks of the Active block to have the style load correctly. You can look at the Attribute block `hasFilterOptions` for inspiration.

roykho replied:
@dinhtungdu - ok I've pushed up a fix for this.

roykho committed:
Hide with iAPI

roykho committed:
Fix linting and e2e

roykho committed:
Merge branch 'trunk' into product-filters/active-block-new-structure

dinhtungdu reviewed (approved):
Thanks for addressing all the feedback! Great work! :shipit:

dinhtungdu committed:
fix: add hidden tag serverside

roykho commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterActive.php:
```diff
@@ -56,18 +56,20 @@ protected function render( $attributes, $content, $block ) {
 		);
 
+		if ( empty( $active_filters ) ) {
+			$wrapper_attributes['hidden'] = true;
```
For my own information, how come this is needed? I thought the iAPI takes care of this? @dinhtungdu

dinhtungdu replied:
@roykho It will, but not the runtime version we're using. We have to manually process the directive for now until we switch to a newer version.
